### DISCUSSION TITLE
can't get react-testing library to work with valtio

### DISCUSSION THREAD

**Discussion by giladv** [pmndrs/valtio#293] [Q&A]
not sure what am i doing wrong but normal code i usually use don't work with valtio:

```
import { render, act } from '@testing-library/react';
...
..

it('should render toasts', () => {

        function Test({ store }) {
		const snap = useSnapshot(store);
		return snap.toasts.map((toast, i) => <div key={i}/>);
	}

	const store = proxy({
		toasts: [],
	});

        function pop(toast){
	        store.toasts.push(ref(toast));
        }

	render(<Test store={store} />);

	act(() => {
		store.pop({});
	});
});

```

this gets me the act error:
```
   Warning: An update to Tst inside a test was not wrapped in act(...).

      When testing, code that causes React state updates should be wrapped into act(...):

      act(() => {
        /* fire events that update state */
      });
      /* assert on the output */

      This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
```

UPDATE: { sync: true } fixes the situation.

**Comment by giladv** [comment]
another craziness with react-testing-library-
the following code will log:
```
- []
- [{}]
- [undefined]
- []
```

```
it('...', () => {

	const store = proxy({
		list: []
	})

	function Test(){
		const snap = useSnapshot(store, {sync: true});
		console.log(snap.list)

		return null;
	}

	render(<Test/>);
        store.list.push(ref({}))
        store.list.splice(0, 1)
});

```

wrapping the splice with ``act`` fixes the issue (which shouldn't be happening regardless).
still, in my use case the splice is happening internally after a setTimeout, which means i cant wrap it with ``act`` so i can't make it work at all unless i mock the timers with jest. 

either way, the entire testing process feels weird.

**Reply by giladv** [reply]
UPDATE: the ``splice`` issue happens on all envs. i managed to reproduce it here.

https://codesandbox.io/s/valtio-simple-counter-forked-z5gyh?file=/src/App.js:77-304

just look at the console.. 
seems to involve around setTimeout + ``{sync: true}``

**Comment by dai-shi** [comment]
Right, the notification is not sync by default. I wonder how we could resolve it without `{ sync: true }`.

**Reply by giladv** [reply]
i believe this is the biggest issue:
https://codesandbox.io/s/valtio-simple-counter-forked-z5gyh?file=/src/App.js:77-304

**Reply by dai-shi** [reply]
Well, it's the correct behavior, because splice has two mutations internally, right?
Same thing can happen with
```js
state.count++
state.count++
```
instead of
```js
state.count += 2
```

**Reply by giladv** [reply]
if JS treats removing a single item from an array with splice as 2 mutations then i believe it should be augmented out somehow.

imagine a component mapping an array to some JSX - it creates an error and it means users are expected to filter arrays from undefined values. 
on mobx for instance, this never happens and it updates sync.

**Reply by dai-shi** [reply]
that's how we batch two mutations into one notification. `{ sync: true }` is a workaround for edge cases.
does mobx trigger two updates for `state.count++; state.count++` and only one for `splice()`? then, implementation-wise, it should be very different. (which is a very interesting finding.)

back to the original question, we need a way to work with `act()` somehow.
Does this happen to work?
```js
  act(() => {
    store.pop({})
    jest.runAllTicks()
  })
```

**Reply by giladv** [reply]
regarding mobx - here's an example
https://codesandbox.io/s/relaxed-moore-rjur8?file=/src/index.js:45-307
if i recall, the way they do it they override common methods.

**Reply by dai-shi** [reply]
> if i recall, the way they do it they override common methods.

I see. Interesting. It's different philosophy. The similar reason why we don't support Set/Map.
Thanks for the remark.

**Comment by giladv** [comment]
hi, @dai-shi, thinking of revisiting valtio again. was this issue solved on newer versions, or updates are still async?

**Reply by dai-shi** [reply]
this is not an issue, but a design choice. we don't override JS methods.
